---
title: "Deconvolution"
output:
  html_document:
    df_print: paged
output.width: "75%"
---

# Library

```{r}
library(Matrix)
library(ggplot2)
library(tidyverse)
library(dplyr)
library(patchwork)


library(STdeconvolve)
library(SpatialExperiment)
library(lemur)
library(igraph)
library(qs)

# library(spacexr)
library(purrr)
library(gridExtra)

# library(ggcorrplot)
# library(imager)
```

```{r}
source("../../src/util.R")
source("../../config/config.R")
```

# Load data

```{r}
set.seed(5)
```

```{r}
dir <-  get.std.path()
spe <- load.data(dir)

pos <- spatialCoords(spe)
pos <- pos[, c('x','y')]
pos <- as.matrix(pos)

res.dir <- get.res.dir()

# load reference data set
ref <- qread(file.path(res.dir, "SCE/sce.qs"))

ctypes <- ref$cell.group.7
table(ctypes)
which(unique(ctypes) == target.cell )

ref.counts.all <- t(counts(ref)) # ctype x gene
```

# Deconvolution

```{r}
target.cell <- "SynFib" 
target.cell <- "Plasma Cells"
target.cell <- "Endothelial Cells"
```

## STD-poisson

```{r}
res.stdp <- read.csv(file.path(res.dir, "deconvolution/std-poisson/deconProp.tsv"), sep = "\t", row.names = 1, header = TRUE)
deconProp.stdp <- res.stdp %>% as.matrix()
deconProp.stdp <- deconProp.stdp / rowSums(deconProp.stdp)

feature.stdp <- read.csv(file.path(res.dir, "deconvolution/std-poisson/expected-features.tsv"), sep = "\t", row.names = 1, header = TRUE)
feature.stdp <- read.csv(file.path(res.dir, "deconvolution/std-poisson/contributions_gene_type.tsv"), sep = "\t", row.names = 1, header = TRUE)
deconGexp.stdp <- feature.stdp %>% as.matrix()
mode(deconProp.stdp) <- "numeric"
```

```{r}
# intercect of genes between spatial and scRNA
genes.sel <- rownames(deconGexp.stdp)
ref.counts <- ref.counts.all[ , genes.sel]

# gene profiles
dim(ref.counts)
mean.na.rm <- partial(mean, na.rm = TRUE)
rownames(ref.counts) <- ctypes

# drop ECs with VWF/pecam1 exceed the range in spatial data
marker.1 <- 'VWF'
marker.2 <- 'PECAM1'
marker.1.max <- max(counts(spe)[marker.1,])
marker.2.max <- max(counts(spe)[marker.2,])
marker.1.max

ref.counts <- as.matrix(ref.counts) %>%
            as.data.frame() %>%
            mutate(ctypes = ctypes) %>% 
            filter(!(!!sym(marker.1) > marker.1.max & ctypes == "Endothelial Cells"))

"PECAM1" %in% colnames(ref.counts)
dim(ref.counts)

ref.gexp <- aggregate(.~ ctypes, data = ref.counts, FUN = mean.na.rm) %>% as.matrix()

ctypes.unique <- ref.gexp[,1]
cell.idx <- which(ctypes.unique == target.cell)
ref.gexp <- ref.gexp[, -1]
mode(ref.gexp) <- "numeric"

# fix typo
ctypes.unique[which(ctypes.unique == "Meyloid")] <- "Myeloid"
```

```{r}
colnames(deconProp.stdp) 
```

```{r}
corMtx_beta <- getCorrMtx(# the deconvolved cell-type `beta` (celltypes x genes)
                          m1 = t(as.matrix(deconGexp.stdp)),
                          # the reference `beta` (celltypes x genes)
                          m2 = as.matrix(ref.gexp),
                          # "b" = comparing beta matrices, "t" for thetas
                          type = "b")
## row and column names need to be characters
rownames(corMtx_beta) <- paste0("decon_", seq(nrow(corMtx_beta)))
colnames(corMtx_beta) <- ctypes.unique

corMtx_beta_paired <- corMtx_beta
colnames(corMtx_beta_paired) <- paste0(seq(length(ctypes.unique)), ":", ctypes.unique)

## Order the cell-types rows based on best match (highest correlation) with each community.
## Cannot have more rows than columns for this pairing, so transpose
pairs <- lsatPairs(t(corMtx_beta_paired))
m <- t(corMtx_beta_paired)[pairs$rowix, pairs$colsix]
corr.plot.std <- correlationPlot(mat = t(m),
                colLabs = "Deconvolved cell-types",
                rowLabs = "Ground truth cell-types",
                title = "PCCs re cell type signatures - Berglund", annotation = T) +
  ## this function returns a `ggplot2` object, so can add additional aesthetics
ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90, vjust = 0, size = 6),
                  axis.text.y = ggplot2::element_text(size = 8))

# colnames(corMtx_beta) <- rownames(ref.gexp)
rownames(corMtx_beta)
colnames(corMtx_beta)
matched.topic <- which.max(corMtx_beta[, cell.idx])
matched.topic
corr.plot.std
```

```{r}
plot.stdp <- vizTopic2(theta = deconProp.stdp, pos = pos, topic = matched.topic, plotTitle = paste('Berglund', matched.topic, target.cell, sep = " - "),
         size = 1, stroke = 0, alpha = 0.5,
         low = "white",
         high = "red")
plot.stdp
```

## STDeconvolve

```{r}
# select the LDA model of interest and get the beta (cell-type transcriptional profiles) and theta (cell-type barcode proportions) matrices.
res.path <- "../../results/deconvolution/STDeconvolve/ldas.qs"
if (!file.exists(res.path)) {
  source("../../scripts/deconvolution/STDeconvolve.R")
}
ldas <- qs::qread(res.path)
ldas$kneedOptK
prior.k <- 7

# ldas$kneedOptK
optLDA <- optimalModel(models = ldas, opt = prior.k)
results <- getBetaTheta(optLDA, perc.filt = 0.05, betaScale = 1000)
deconProp.stdeconv <- results$theta
deconGexp <- results$beta
# qs::qsave(deconProp, )
dim(deconProp.stdeconv)
```

```{r}
## visualize deconvolved cell-type proportions
# vizAllTopics(deconProp, pos[rownames(deconProp),], r=55, lwd=0)	  

## interpret these cell-types based on their gene expression
# topGenes(deconGexp)
## top genes based on log 2 (fold change) for each cell-type
get.marker.genes <-  function(i, n = 10, gexp = deconGexp) {
  head(sort(log2(gexp[i,]/colMeans(gexp[-i,])), decreasing=TRUE), n = n)
}
```

### Annotation strategy

```{r}
genes.sel <- colnames(deconGexp)
ref.counts <- ref.counts.all[ ,genes.sel]

# gene profiles
dim(ref.counts)
mean.na.rm <- partial(mean, na.rm = TRUE)
rownames(ref.counts) <- ctypes
ref.counts <- as.matrix(ref.counts) %>%
            as.data.frame() %>%
            mutate(ctypes = ctypes)
  
ref.gexp <- aggregate(.~ ctypes, data = ref.counts, FUN = mean.na.rm) %>% as.matrix()

ctypes.unique <- ref.gexp[,1]
cell.idx <- which(ctypes.unique == target.cell)
ref.gexp <- ref.gexp[, -1]
mode(ref.gexp) <- "numeric"

ctypes.unique
rownames(ref.gexp)
```

```{r}
corMtx_beta <- getCorrMtx(# the deconvolved cell-type `beta` (celltypes x genes)
                          m1 = as.matrix(deconGexp),
                          # the reference `beta` (celltypes x genes)
                          m2 = as.matrix(ref.gexp),
                          # "b" = comparing beta matrices, "t" for thetas
                          type = "b")
```

```{r}
## row and column names need to be characters
rownames(corMtx_beta) <- paste0("decon_", seq(nrow(corMtx_beta)))

corr.plot.stdeconv <- correlationPlot2(mat = corMtx_beta,
                colTicks = ctypes.unique, 
                colLabs = "Deconvolved cell-types",
                rowLabs = "Ground truth cell-types",
                title = "PCCs re cell type signatures - STDeconvolve", annotation = T, annotation_size = 4) +
  ## this function returns a `ggplot2` object, so can add additional aesthetics
ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90, vjust = 0, size = 6),
                  axis.text.y = ggplot2::element_text(size = 8))

# colnames(corMtx_beta) <- rownames(ref.gexp)
rownames(corMtx_beta)
colnames(corMtx_beta)
matched.topic <- which.max(corMtx_beta[, cell.idx])
matched.topic
corMtx_beta[matched.topic, cell.idx]
corr.plot.stdeconv
```

```{r}
plot.stdeconv <- vizTopic2(theta = deconProp.stdeconv, pos = pos, topic = matched.topic, plotTitle = paste('STDeconv', matched.topic, target.cell, sep = " - "),
         size = 1, stroke = 0, alpha = 0.5,
         low = "white",
         high = "red")
```

## RCTD

```{r}
# res.rctd <- readRDS(paste0(ref.dir, "SHK166_RCTD.rds"))
res.rctd <- qs::qread(file.path(res.dir, "deconvolution/RCTD/myRCTD.qs"))
res.rctd <- res.rctd@results
ctypes <- colnames(res.rctd$norm_weights)
ctypes
deconProp.rctd <- res.rctd$norm_weights %>% as.matrix()
cell.idx.rctd <- which(colnames(deconProp.rctd) == target.cell )
cell.idx.rctd
```

```{r}
# pos
# deconProp
head(deconProp.rctd)
# deconProp.rctd
min(deconProp.rctd[,cell.idx.rctd])
max(deconProp.rctd[,cell.idx.rctd])

colnames(deconProp.rctd) <- seq_len(ncol(deconProp.rctd))
plot.rctd <- vizTopic2(theta = deconProp.rctd, pos = pos, topic = cell.idx.rctd, plotTitle = paste0("RCTD - ", target.cell),
         size = 1, stroke = 0, alpha = 0.5,
         low = "white",
         high = "red")
plot.rctd
```

## Cell2location

```{r}
res.c2l <- read.csv(file.path(res.dir, "deconvolution/cell2loc/deconProp.csv"), row.names = 1)
deconProp.c2l <- res.c2l%>% as.matrix()
colnames(deconProp.c2l) <- ctypes.unique
ctypes.unique
cell.idx.c2l <- which(colnames(deconProp.c2l) == target.cell )

colnames(deconProp.c2l) 
cell.idx.c2l
```

```{r}
colnames(deconProp.c2l) <- seq_len(ncol(deconProp.c2l))
rownames(deconProp.c2l) <- rownames(deconProp.stdp)
sum(deconProp.c2l)

plot.c2l <- vizTopic2(theta = deconProp.c2l, pos = pos, topic = cell.idx.c2l, plotTitle = paste0("cell2loc - ", target.cell),
         size = 1, stroke = 0, alpha = 0.5,
         low = "white",
         high = "red")

plot.c2l
```

## Comparison

### Average proportions Barplot

```{r}
avg.prop.bp <- function(deconProp, type) {
  avg.prop <- colSums(deconProp) / nrow(deconProp)
  
  bp <- barplot(
    avg.prop,
    names.arg = ctypes,
    # names.arg = seq(length(avg.prop)),
    las = 2,  # Rotate x-axis labels to be vertical
    col = "lightblue",
    main = paste0("Average Proportions by Cell Type - ", type)
  )
  return(bp)
}

plot1 <- avg.prop.bp(deconProp.stdeconv, "STDeconv") # STDeconv
plot2 <- avg.prop.bp(deconProp.rctd, "RCTD") # STDeconv
plot3 <- avg.prop.bp(deconProp.c2l, "Cell2Loc") # STDeconv
plot4 <- avg.prop.bp(deconProp.stdp, "Berglund") # STDeconv

# plots.grid <- grid.arrange(plot1, plot2, plot3, plot4, ncol = 2)

# ggsave(filename = "Average Proportions.svg", plot = plots.grid, device = "svg", width = 8, height = 6)
# plots.grid
```

### Visualization of a certain cell type

```{r}
filename <-  paste0("Visualization of ", target.cell, ".svg")

plots.grid <- wrap_plots(plot.stdeconv, plot.rctd, plot.c2l, plot.stdp)
# plots.grid <- wrap_plots(plot.stdeconv, plot.rctd, plot.stdp)

ggsave(filename = file.path(res.dir, "plots", filename), plot = plots.grid, width = 12, height = 6)

plots.grid
```

```         
```

### Agreement on marker genes

```{r}
pred.mg <- get.marker.genes(matched.topic, n = 100)
true.mg <- get.marker.genes(matched.topic, n = 100, gexp = ref.gexp)

diff.genes <- intersect(names(pred.mg), names(true.mg))

```

## Store results

```{r}
filename <-  file.path(res.dir, "deconvolution","deconProp_best.qs")
qsave(deconProp.stdp, filename)

filename <-  file.path(res.dir, "plots"," PCCs.pdf")
plots.grid <- wrap_plots(corr.plot.stdeconv, corr.plot.std)
ggsave(filename = filename, plot = plots.grid, device = "svg", width = 12, height = 6)

```

# Annotate ECs manually

```{r}
marker.sel <- "PECAM1"
threshold <- 5

marker.sel <- "VWF"
threshold <- 2

```

```{r}
marker.genes <- qread(file.path(res.dir, "markers.qs"))
table(marker.genes$cluster)

markers.ec <- marker.genes %>% 
              group_by(cluster) %>%
              top_n(n = 100, wt = avg_log2FC) %>% 
              filter(cluster == "Endothelial Cells" & pct.1 > 0.25 & pct.2 >= 0) 
head(markers.ec, n = 10)


markers.synFib <- marker.genes %>% 
              group_by(cluster) %>%
              top_n(n = 100, wt = avg_log2FC) %>% 
              filter(cluster == "SynFib" & pct.1 > 0.25 & pct.2 >= 0) 
head(markers.synFib, n = 10)
"CD90" %in% markers.synFib

# idx.marker.sel <- which(markers.ec$gene == marker.sel)
# markers.ec[idx.marker.sel, ]

gexp.target <- counts(spe)[marker.sel, ]
marker.sel %in% rownames(marker.genes)
table(gexp.target)
```

```{r}
# find threshold by scRNA-seq barplot for ECs and others
df <- data.frame(
  count = counts(ref)[marker.sel, ],
  ctypes = ref$cell.group.7
)

max(df$count)
p <- as_tibble(df) %>% 
    filter(ctypes == "Endothelial Cells") %>% 
    ggplot(aes(x = count,  fill = ctypes)) +
    geom_bar() +
    facet_wrap(~ ctypes, scales = "free_y") +
    theme_minimal() +
    labs(title = "Counts of VWF in Endothelial Cells and Other Cells",
         x = "VWF",
         y = "Count")
p


p <- as_tibble(df) %>% 
    filter(ctypes != "Endothelial Cells") %>% 
    ggplot(aes(x = count,  fill = ctypes)) +
    geom_bar() +
    facet_wrap(~ ctypes, scales = "free_y") +
    theme_minimal() +
    labs(title = "Counts of VWF in Endothelial Cells and Other Cells",
         x = "VWF",
         y = "Count")
p

p <- as_tibble(df) %>% 
    filter(count > 0) %>% 
    ggplot(aes(x = count,  fill = ctypes)) +
    geom_bar() +
    facet_wrap(~ ctypes, scales = "free_y") +
    theme_minimal() +
    labs(title = "Counts of VWF in Endothelial Cells and Other Cells",
         x = "VWF",
         y = "Count")
p

p <- as_tibble(df) %>% 
    filter(count > quantile(df$count, 0.05) & count < quantile(df$count, 0.95)) %>% 
    ggplot(aes(x = count,  fill = ctypes)) +
    geom_bar() +
    facet_wrap(~ ctypes, scales = "free_y") +
    theme_minimal() +
    labs(title = "Counts of VWF in Endothelial Cells and Other Cells",
         x = "VWF",
         y = "Count")
p

p <- as_tibble(df) %>% 
    filter(count < quantile(df$count, 0.95)) %>% 
    ggplot(aes(x = count,  fill = ctypes)) +
    geom_bar() +
    facet_wrap(~ ctypes, scales = "free_y") +
    theme_minimal() +
    labs(title = "Counts of VWF in Endothelial Cells and Other Cells",
         x = "VWF",
         y = "Count")
p

gexp.target <- counts(spe)[marker.sel, ]
# marker.sel %in% rownames(marker.genes)
table(gexp.target)

```

```{r}
# Create the bar plot using ggplot2
plots <- as_tibble(gexp.target) %>% 
  ggplot(aes(x = value)) +
    geom_bar(fill = "skyblue") +
    theme_minimal() +
    labs(title = paste0("Frequency of ", marker.sel, " across All Spots"),
         x = "UMI",
         y = "Frequency")

filename <- paste0("Bar plot of Frequency of ", marker.sel, ".svg")
ggsave(filename = file.path(res.dir, "plots", filename), plot = plots, device = "svg", width = 10, height = 8)

filename <- paste0("Bar plot of Frequency of ", marker.sel, ".png")
ggsave(filename = file.path(res.dir, "plots", filename), plot = plots, device = "png", width = 10, height = 8, dpi = 500)

is.ec.spot <- (gexp.target > threshold)
dim(is.ec.spot)

filename <- paste0("ECs_manually_", marker.sel, ".qs")
qsave(as.data.frame(is.ec.spot), file.path(res.dir, "deconvolution", filename))
plots
```

```{r}
method <- paste0("Manually by marker ", marker.sel)

gene_of_interest <- "THY1"
gene_of_interest <- marker.sel
cluster <- "Endothelial Cells"
  
groups <- rep("others", ncol(spe))
groups[unlist(is.ec.spot)] <- cluster

gcnt <- counts(spe)[gene_of_interest, ]
strokes <- ifelse(groups == cluster, 0.5, 0)

# Create ggplot object
plots <- as_tibble(spatialCoords(spe)) %>%
  mutate(gcnt = gcnt, groups = groups, strokes = strokes) %>%
  ggplot(aes(x = x, y = y, shape = groups, fill = gcnt)) +
  geom_point(aes(stroke = strokes), size = 2, alpha = 0.7, color = "black") +
  scale_shape_manual(values = c(24, 21)) +
  scale_fill_gradient(low = "white", high = "red") +
  labs(title = paste0(gene_of_interest, " vs ", "ECs from ", method),
       x = "pxl_row_in_fullres",
       y = "pxl_col_in_fullres",
       fill = paste0(gene_of_interest, " Level"))

filename <- paste0(gene_of_interest, " vs ", "ECs from ", method,  ".svg")
ggsave(filename = file.path(res.dir, "plots", filename), plot = plots, device = "svg", width = 10, height = 8)

filename <- paste0(gene_of_interest, " vs ", "ECs from ", method,  ".png")
ggsave(filename = file.path(res.dir, "plots", filename), plot = plots, device = "png", width = 10, height = 8, dpi = 500)

plots
```

# THY1 Plot

```{r}
gene_of_interest <-  "THY1"
threshold <- 0.6
  
# Assuming you have defined `threshold`, `deconProp.rctd`, `deconProp.stdeconv`, and other necessary variables

# Define topic targets and methods
targets <- list(
  list(topic.target = 2, method = "RCTD", deconProp = deconProp.rctd),
  list(topic.target = 1, method = "STDeonv", deconProp = deconProp.stdeconv),
  list(topic.target = 3, method = "Berglund", deconProp = deconProp.stdp)
)

# Initialize empty list to store plots
plot_list <- list()

# Iterate over each target
for (target in targets) {
  topic.target <- target$topic.target
  method <- target$method
  deconProp <- target$deconProp
  
  # Compute pos.target.mask (assuming threshold is predefined)
  pos.target.mask <- deconProp[, topic.target] > threshold
  groups <- rep("others", ncol(spe))
  groups[pos.target.mask] <- "targets"
  gcnt <- counts(spe)[gene_of_interest, ]
  strokes <- ifelse(groups == "targets", 0.5, 0)
  
  # Create ggplot object
  plot <- as_tibble(spatialCoords(spe)) %>%
    mutate(gcnt = gcnt, groups = groups, strokes = strokes) %>%
    ggplot(aes(x = x, y = y, shape = groups, fill = gcnt)) +
    geom_point(aes(stroke = strokes), size = 2, alpha = 0.7, color = "black") +
    scale_shape_manual(values = c(21, 24)) +
    scale_fill_gradient(low = "white", high = "red") +
    labs(title = paste0(gene_of_interest, " vs ", "ECs from ", method, " - threshold - ", threshold),
         x = "pxl_row_in_fullres",
         y = "pxl_col_in_fullres",
         fill = "THY1 Level")
  # Store plot object in list
  plot_list[[length(plot_list) + 1]] <- plot
}
  
plots <- wrap_plots(plot_list, ncol = 2)

filename <- paste0(" Dist of ", "threshold - ", threshold, gene_of_interest, ".svg")
ggsave(filename = file.path(res.dir, "plots", filename), plot = plots, device = "svg", width = 10, height = 8)
  
plots
```

# Utils

```{r}
hist(deconProp[, matched.topic], breaks = 5)
```

# Misc

# Reference

-   Tutorial from Prof. Jean Fan, <https://jef.works/blog/2023/05/29/stdeconvolve-breast-cancer/>

-   STdeconvolve Vignettes, <https://www.bioconductor.org/packages/release/bioc/vignettes/STdeconvolve/inst/doc/vignette.html#spatialexperiment-inputs>
